--- host.c.orig	2018-05-19 17:38:16.000000000 +0300
+++ host.c	2018-05-19 17:43:33.000000000 +0300
@@ -198,9 +198,19 @@
                 (CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE & cmd_q_props)
                 ? "yes" : "no");
 
-    /*
-     * type here your OpenCL API call
-     */
+    /* Solution for the option 2 */
+    printf("\t\tEndianness: ");
+    cl_bool endianness;
+    rv = clGetDeviceInfo(device, CL_DEVICE_ENDIAN_LITTLE,
+           sizeof(cl_bool), &endianness, NULL);
+    if(CL_SUCCESS != rv)
+    {
+       print_cl_error("failed - ", rv, stderr);
+    }
+    else
+    {
+       printf("%s\n", endianness ? "little" : "big");
+    }
 
     // get information on extensions
     size_t ext_size;
